### PR TITLE
Add a api to reset and extract RNG Seed

### PR DESCRIPTION
**Pull request by JackCaoG**
This pr is the pytorch/xla side of fix for https://github.com/pytorch/xla/issues/4169.

In this pr we provided a way to check whether a device_data is a IR seed and add an api to get the `device_data` corresponding to the random seed. Remaining part is to add a check in graphInputMatcher https://github.com/pytorch/pytorch/blob/e4a8661ab84022c1bff622c6d2f6e679180b1df5/torch/_dynamo/optimizations/torchxla_integration.py#L36

something similar to 
```
if tensor_id == seed_info_id:
  inp = torch_xla._XLAC._get_rng_seed_as_tensor(str(traced_ivalue.device), True)
```

Whether to `Reset` the IR seed depends on if we do a `mark_step` across different dynamo runs. (mark_step will increment base seed and we want different base seed for every step).

FYI @shunting314 @wconstab

### CODE REVIEW

**Comment by shunting314** [comment]
```
if tensor_id == seed_info_id:
  inp = torch_xla._XLAC._get_rng_seed_as_tensor(str(traced_ivalue.device), True)
```

This should be done at inference/training time rather than compiling time, right?

**Comment by JackCaoG** [comment]
> ```
> if tensor_id == seed_info_id:
>   inp = torch_xla._XLAC._get_rng_seed_as_tensor(str(traced_ivalue.device), True)
> ```
> 
> This should be done at inference/training time rather than compiling time, right?

right, we want to use the new seed for every execution.

File: torch_xla/csrc/init_python_bindings.cpp [hunk 1]
```diff
@@ -1498,27 +1500,39 @@ void InitXlaModuleBindings(py::module m) {
...
-
-            // add tensor_id after we make sure the handle does not exist yet.
-            tensor_ids.push_back(infoptr->tensor_id);
+            auto* infoptr =
+                static_cast<torch::lazy::LazyGraphExecutor::DeviceDataInfo*>(
+                    backend_data->info());
+            if (infoptr) {
+              tensor_ids.push_back(infoptr->tensor_id);
+            } else {
+              // TODO(JackCaoG): Make sure this device data is actually seed.
```
Comment (shunting314) [inline torch_xla/csrc/init_python_bindings.cpp:1516]: yep, I'm also worried if there would be DeviceData node other than seed that has null InfoPtr

File: torch_xla/csrc/tensor.cpp [hunk 2]
```diff
@@ -367,6 +367,27 @@ class XLATensor::DeviceContextArena {
     devctx->seed_ir_value = torch::lazy::Value();
   }
 
+  torch::lazy::BackendDataPtr GetRngSeedData(
+      const torch::lazy::BackendDevice& device, bool reset) {
```
Comment (shunting314) [inline torch_xla/csrc/tensor.cpp:371]: Can you explain a bit how this 'reset' arg works? In dynamo if we want to replace the seed placeholder with the real seed, should we call this method with reset to be true?

**Review by shunting314 — APPROVED** [review]
Thank you for adding this. Adding comments for my understanding.

File: torch_xla/csrc/init_python_bindings.cpp [hunk 3]
```diff
@@ -58,6 +58,8 @@
 namespace torch_xla {
 namespace {
 
+static int64_t seed_info_id = -127389;
```
[the remark this answers is not in the document]
Comment (JackCaoG) [reply torch_xla/csrc/init_python_bindings.cpp:61]: nope, I just randomly pick a negative number since regular tensor ID can not be negative. (scalar tensor id would be -1)

File: torch_xla/csrc/tensor.cpp (hunk 2 above)
Comment (JackCaoG) [reply torch_xla/csrc/tensor.cpp:371]: if you pass `reset == false`, and keep calling `GetRngSeedData` it will always return you the same seed data. If you pass `reset==true`, it will give you a new seed.

However `mark_step` will also increment the base seed, and I believe we call `mark_step` for every `optim_mode` now. If that's the case we don't need to `reset` the seed in this API.

File: torch_xla/csrc/init_python_bindings.cpp (hunk 1 above)
Comment (JackCaoG) [reply torch_xla/csrc/init_python_bindings.cpp:1516]: I will add the check in a following pr.
